internal/state/mailbox_fetch.go

```diff
@@ -111,12 +112,26 @@ func (m *Mailbox) Fetch(ctx context.Context, seq *proto.SequenceSet, attributes
 		var literal []byte
 
 		if needsLiteral {
-			l, err := m.state.getLiteral(msg.ID.InternalID)
-			if err != nil {
-				return err
-			}
+			l, firstErr := m.state.getLiteral(msg.ID.InternalID)
+			if firstErr != nil {
+				logrus.Debugf("Failed load %v from store, attempting to download from connector", msg.ID.InternalID.ShortID())
+
+				connectorLiteral, err := m.state.user.GetRemote().GetMessageLiteral(ctx, message.RemoteID)
+				if err != nil {
+					logrus.Errorf("Failed to download message from connector: %v", err)
+					return firstErr
+				}
 
-			literal = l
+				if err := m.state.user.GetStore().Set(msg.ID.InternalID, connectorLiteral); err != nil {
+					logrus.Errorf("Failed to store download message from connector: %v", err)
+					return firstErr
+				}
+
+				logrus.Debugf("Message %v downloaded and stored ", msg.ID.InternalID.ShortID())
+				literal = connectorLiteral
+			} else {
+				literal = l
+			}
```

Comment:
Just a thought, could `state.getLiteral` transparently fetch/save the message if necessary, thereby hiding the complexity from the mailbox? This block of code is duplicated in mailbox_fetch.go and mailbox_search.go when it could probably be done once in `state.getLiteral` instead.

Reply:
Good point, fixed.